File: frontend/src/api/tag.ts
import {
  CreateTagPayload,
  CreateTagResponse,
  DeleteTagPayload,
  DeleteTagResponse,
  UpdateTagPayload,
  UpdateTagResponse
} from 'generated/models'
import { RequestAPI } from './request'

// All endpoints under the '/tags' prefix
export class TagAPI {
  private req: RequestAPI

  constructor() {
    this.req = new RequestAPI('/tags')
  }

  async createTag(payload: CreateTagPayload): Promise<CreateTagResponse> {
    return this.req.post('/', payload)
  }

  async deleteTag(payload: DeleteTagPayload): Promise<DeleteTagResponse> {
    return this.req.delete(`/${payload.id}`)
  }

  async editTag(payload: UpdateTagPayload): Promise<UpdateTagResponse> {
    return this.req.put('/', payload)
  }
}

File: backend/services/tag/tag.go
package services

import (
	"github.com/EmilyOng/cvwo/backend/db"
	"github.com/EmilyOng/cvwo/backend/models"
)

func CreateTag(payload models.CreateTagPayload) models.CreateTagResponse {
	tag := models.Tag{Name: payload.Name, Color: payload.Color, BoardID: &payload.BoardID}
	result := db.DB.Create(&tag)
	return models.CreateTagResponse{
		Response: models.Response{Error: result.Error},
		Tag:      models.TagPrimitive{ID: tag.ID, Name: tag.Name, Color: tag.Color, BoardID: tag.BoardID},
	}
}

func DeleteTag(payload models.DeleteTagPayload) models.DeleteTagResponse {
	tag := models.Tag{ID: payload.ID}
	result := db.DB.Model(&tag).Preload("Tasks").Find(&tag)
	if result.Error != nil {
		return models.DeleteTagResponse{
			Response: models.Response{Error: result.Error},
		}
	}
	err := db.DB.Model(&tag).Association("Tasks").Delete(&tag.Tasks)
	if err != nil {
		return models.DeleteTagResponse{
			Response: models.Response{Error: err},
		}
	}
	result = db.DB.Delete(&tag)
	return models.DeleteTagResponse{
		Response: models.Response{Error: result.Error},
	}
}

func UpdateTag(payload models.UpdateTagPayload) models.UpdateTagResponse {
	tag := models.Tag{ID: payload.ID, Name: payload.Name, BoardID: &payload.BoardID, Color: payload.Color}
	result := db.DB.Model(&models.Tag{ID: tag.ID}).Save(&tag)
	return models.UpdateTagResponse{
		Response: models.Response{Error: result.Error},
		Tag:      models.TagPrimitive{ID: tag.ID, Name: tag.Name, BoardID: tag.BoardID, Color: tag.Color},
	}
}
